### ISSUE TITLE
CMake and finding libcurl error in 4.9.2

### ISSUE THREAD

**Issue by KBentley57** [Unidata/netcdf-c#3031]
To report a non-security related issue, please provide:

* the version of the software with which you are encountering an issue

 4.9.2

* environmental information (i.e. Operating System, compiler info, java version, python version, etc.)

rocky linux 8, gcc 13, cmake 3.26.5

* a description of the issue with the steps needed to reproduce it

I've built netcdf-c on a host of many different machines, and a few have always had trouble "finding" curl.  I was curious as to why, since I'd given the cmake practically every hint I could think of.  After poking around in there, I think I've got a suggestion that will help.

It turns out it was finding it.  but the check for CURL_LIBRARY OR CURL_LIBRARIES around line 1000 was failing.  I don't know why, or if something in the CMake module for findCURL has changed in later versions, but the 'modern' curl config has existed for a good bit of time now.  The call to find_package(CURL) will first search the cmake module path for the CURLConfig.cmake file, and if found, it will create the target CURL::libcurl.  I'd suggest looking to see if this target exists directly, rather than relying on result variables that may now be unused somehow.  Any combination of checking for this too alongside the other variables would probably work too. I changed those lines to be 

```
find_package(CURL)
if(TARGET CURL::libcurl)
    message(STATUS "CURL: FOUND")
    set(FOUND_CURL TRUE)
else()
    message(STATUS "CURL: NOT FOUND")
    set(FOUND_CURL FALSE)
endif()
```

There is an associated change in the liblib CMakeLists.txt around line 125, to add `CURL::libcurl` to the `TLL_LIBS` variable.

and it was able to pass that section just fine.  Obviously the message could be taken out, but it's nice to know where that kind of subtle error might come one.  

If you have a general question about the software, please view our [Suggested Support Process](https://www.unidata.ucar.edu/support/#process).

**Comment by WardF** [comment]
THe current version (v4.9.3-rc1) checks for `libcurl` as shown [here](https://github.com/Unidata/netcdf-c/blob/b84fcd7e30f72c6da24b50f55326364e87b2ac11/cmake/dependencies.cmake#L237).  It looks to achieve the same thing you propose here, but perhaps there is a functional difference I'm not seeing.  Does the change at the link above look good, resolve the issue?

**Comment by KBentley57** [comment]
@WardF I'm going to try that out momentarily, I'll report back.

**Comment by KBentley57** [comment]
Just to add more information here,  I tried it on another PC with rocky 8, and using 4.9.2 it displayed the same behavior, but it does indeed find curl.  If I look in the CMakeCache I can see that the variables are defined and populated.
```
CURL_DIR:PATH=/data1/SupportPackages/lib/cmake/CURL
CURL_INCLUDE_DIR:PATH=/usr/include
CURL_LIBRARY_DEBUG:FILEPATH=CURL_LIBRARY_DEBUG-NOTFOUND
CURL_LIBRARY_RELEASE:FILEPATH=/usr/lib64/libcurl.so
````
Furthermore, that pkg-config output `pkg-config --libs libcurl` shows 
```
-L/data1/SupportPackages/lib -lcurl
```

I was checking through the FindCURL.cmake provided by /usr/share/cmake3/FindCURL.cmake on my distro, and I saw that there's a variable `CURL_NO_CURL_CMAKE` that can be set to `ON` to use a 'legacy' method of find package, instead of looking for the cmake-exported configuration file that curl generates and places in ${CMAKE_INSTALL_PREFIX}/lib/cmake.  I set this right above the lines in the 4.9.2 CMakeLists.txt and it was able to find curl just fine.  
```
986     # See if we have libcurl
987     set(CURL_NO_CURL_CMAKE ON)
988     FIND_PACKAGE(CURL)
989     ADD_DEFINITIONS(-DCURL_STATICLIB=1)
990     INCLUDE_DIRECTORIES(${CURL_INCLUDE_DIRS})
```
It look like it at least imported the targets correctly, which is why my first fix worked, skipping the result variables that are obviously set incorrectly.  So, perhaps this isn't a netcdf  issue at all, rather an issue with the way curl (or the version I'm using) generates it's config file that the cmake module tries to find, and use, first.  

I'll try to find more info and post back if I nail down the exact cause.  However, I'm getting a few other issues on the build that make me suspect something with the cmake package, I get a different configuration error or two like this one, that seems dubious.

```
CMake Error at CMakeLists.txt:2510 (ADD_MAKEDIST):
  Unknown CMake command "ADD_MAKEDIST".
```
Thanks,

Kyle

**Comment by KBentley57** [comment]
I was using curl 7.78.0, and decided to pull the latest, 8.10.1.  That seemed to fix the issue I was having related to curl.  I don't think it's worth investigating further, given the odds of someone else having the same combination of versions.  The other error about the "ADD_MAKEDIST" were related to overriding the CMAKE_MODULE_PATH so that I could find other projects inside netcdf.  

I looked for that in the CMake, and it looks like CMAKE_MODULE_PATH is being clobbered on lines 111-113
```
111    #Add custom CMake Module
112    SET(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules/"
113    CACHE INTERNAL "Location of our custom CMake modules.")
```
which may be undesirable if a user passes in a module path with `-DCMAKE_MODULE_PATH=/additional/paths/for/files...`  from the command line.  Would it be worth modifying that section to be 

```
list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules/")
```

instead?  It'll behave the same way, but prevent an error if a user assumes that the module path is modified instead of restated.

Thanks,

Kyle

**Comment by WardF** [comment]
Thanks! We are not using makedist and that can be removed; regarding your suggested change to the module path, that makes sense.  It should have likely been like that from the start, and this is just the first time we've encountered it. Thanks! I'll put a fix up.

**Comment by WardF** [comment]
I see this was already addressed in `v4.9.3-rc1`; I will mark this as closed, then, and we can reopen if needs be! Thanks!
